Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/SyncProgramsJobService.java

```diff
@@ -0,0 +1,164 @@
+package org.dolphinemu.dolphinemu.services;
+
+import android.app.job.JobParameters;
+import android.app.job.JobService;
+import android.content.ContentUris;
+import android.content.Context;
+import android.database.Cursor;
+import android.net.Uri;
+import android.os.AsyncTask;
+import android.os.PersistableBundle;
+import android.support.media.tv.Channel;
+import android.support.media.tv.PreviewProgram;
+import android.support.media.tv.TvContractCompat;
+import android.util.Log;
+
+import org.dolphinemu.dolphinemu.R;
+import org.dolphinemu.dolphinemu.model.GameFile;
+import org.dolphinemu.dolphinemu.ui.platform.Platform;
+import org.dolphinemu.dolphinemu.utils.AppLinkHelper;
+import org.dolphinemu.dolphinemu.utils.TvUtil;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class SyncProgramsJobService extends JobService
+{
+	private static final String TAG = "SyncProgramsJobService";
+
+	private SyncProgramsTask mSyncProgramsTask;
+
+	@Override
+	public boolean onStartJob(final JobParameters jobParameters)
+	{
+		Log.d(TAG, "onStartJob(): " + jobParameters);
+		final long channelId = getChannelId(jobParameters);
+		if (channelId == -1L)
+		{
+			Log.d(TAG, "Failed to find channel");
+			return false;
+		}
+
+		mSyncProgramsTask =
+			new SyncProgramsTask(getApplicationContext())
+			{
+				@Override
+				protected void onPostExecute(Boolean finished)
+				{
+					super.onPostExecute(finished);
+					mSyncProgramsTask = null;
+					jobFinished(jobParameters, !finished);
+				}
+			};
+		mSyncProgramsTask.execute(channelId);
+		return true;
+	}
+
+	@Override
+	public boolean onStopJob(JobParameters jobParameters)
+	{
+		if (mSyncProgramsTask != null)
+		{
+			mSyncProgramsTask.cancel(true);
+		}
+		return true;
+	}
+
+	private long getChannelId(JobParameters jobParameters)
+	{
+		PersistableBundle extras = jobParameters.getExtras();
+		return extras.getLong(TvContractCompat.EXTRA_CHANNEL_ID, -1L);
+	}
+
+	private class SyncProgramsTask extends AsyncTask<Long, Void, Boolean>
```

Comment:
static

Reply:
A static class deriving from a different class? Is that a thing in Java, or did you simply miss the `extends AsyncTask<...>` at the end?

Reply:
@BhaaLseN yes, that is completely valid and encouraged in Java whenever possible to have a static inner class so you don't have access to the outer class. Two positive things to gain, you won't get access to the outer class reference so you can avoid any possible memory leak from that access. The second thing due to the limited access to the outer class you get clean classes (Inner/Outer) that can be separated later if needed into their own java files.

Reply:
Oh, so it's not like a `static class` (cannot create an instance) but more like a visibility/scope feature (reusing the `static` keyword)? It feels strange to create an instance of a class marked as `static`, but it appears that Java has a different meaning for them when they're nested. 👍

Reply:
Fixed. I also went ahead and made the Context that is passed a weak reference.